`st.chat_input` with file uploader

kajarenc opened the pull request:
## Describe your changes

Supports file upload for `st.chat_input`
[Demo](https://file-uploader-in-chat-input.streamlit.app/)

### API

**Enabling and controlling file upload**

Introduced new parameters `accept_file` and `file_type`
```py
st.chat_input(..., accept_file=False|True|"multiple", file_type=None|str|List[str])

# E.g. to accept a single image:
st.chat_input(accept_file=True, file_type=["png", "jpg"])
# Or to accept multiple text files:
st.chat_input(accept_file="multiple", file_type="txt")
```

**Return value**

If accepting file, returns a `Mapping` object of the following form:
```py
message = st.chat_input(accept_file=True)
# where `message` has the following format:
{
    "text": "sample text", 
    "files": [...]
}
```

### Design and Behavior
#### Upload Area
Introduced 
1. an upload button (with help tooltip) to open file chooser

  <img width="219" alt="image" src="https://github.com/user-attachments/assets/21391f10-6b4d-4e90-ba34-acf422b4c150" />

2. Expanded dropzone when dragged target enters **_the browser window_** 

  https://github.com/user-attachments/assets/e138f8eb-cef8-4fd5-a21b-c1571dd41ee3

#### Uploaded File Preview

Added a section above chat input box to display the list of uploaded files
Each file could have 1 of 3 statuses: uploading, error, or uploaded. Hovering on the error icon would show the error message.

![image](https://github.com/user-attachments/assets/534b32d2-b914-4bb5-b33c-412ae0e632ca)

https://github.com/user-attachments/assets/0bc2da1f-d56c-44cd-a960-9b3d91421b5d





## GitHub Issue Link (if applicable)

#7409

## Testing Plan

- Python & JS Unit Tests
- E2E Tests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

kajarenc committed:
initial commit

kajarenc committed:
add frontend scaffolding

kajarenc committed:
add file uploader internals to chat input

kajarenc committed:
useRef + state hack solution

kajarenc committed:
send files to server

kajarenc committed:
remove useRef, rely only on useState

kajarenc committed:
rewrite without using Promise

kajarenc committed:
add functionality for single file upload

kajarenc committed:
allow submitting chat input if only file uploaded without text

kajarenc committed:
Add file_type parameter

kajarenc committed:
Add "multiple" option for accept_file parameter

kajarenc committed:
hide file_uploader if it is configured from backend

kajarenc committed:
resolve merge conflict

kajarenc committed:
try to fix js test to be able to build wheel file

kajarenc committed:
add help parameter back to the public API

kajarenc committed:
remove unused imports after the merge

sfc-gh-pchiu committed:
[WIP] Chat input with file uploder UI (#9543)

[Figma](https://www.figma.com/design/sOLdImJt29L9YYrjlzGaaO/St.chat?node-id=1772-650&node-type=canvas&m=dev)
> Note: We're **not** changing the border radii to be rounded like on
Figma, yet. Keeping the original set up for now.
* use attach_file icon
* fix text input border misalignment; test on text input only widget as
well
* highlight entire chat input area instead of only textarea (which is
the default behavior of [baseui
textarea](https://baseweb.design/components/textarea/), so we need to
[remove](https://github.com/streamlit/streamlit/pull/9543/files#diff-559594d48f19e6ea4c58bfa5cb445c7f4e4596c3865e7e88b7f0af13abf146eaR430)
it)
* margin/padding/gap polish

<img width="810" alt="image"
src="https://github.com/user-attachments/assets/a7049191-0f57-4b68-a198-25d43c440c86">

<img width="758" alt="image"
src="https://github.com/user-attachments/assets/1e1b9111-e796-4971-9a89-c97be5617d88">

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kajarenc committed:
fix value_time for "chat_input"

kajarenc committed:
Add attribute access for chat_input (#9585)

## Describe your changes

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kajarenc committed:
Add draft docstrings (#9601)

## Describe your changes

Add draft docstrings

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kajarenc committed:
fix typing issues, fix overlap issue on frontend

kajarenc committed:
add accept_file and file_type to compute element id

sfc-gh-pchiu committed:
Update icon size for file upload in chat input (#9708)

## Describe your changes

update icon size according to
[design](https://www.figma.com/design/sOLdImJt29L9YYrjlzGaaO/St.chat?node-id=1772-650&node-type=canvas&m=dev)

#### Before 
![Screenshot 2024-10-22 at 11 50
05 AM](https://github.com/user-attachments/assets/015de2d2-9273-4168-a817-ff4a751d89a7)


#### After
![Screenshot 2024-10-22 at 11 48
32 AM](https://github.com/user-attachments/assets/18f4fa7c-b942-4243-b174-71b27ccf3c64)



## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
auto focus on text input after file upload (#9709)

## Describe your changes

After uploading files, we want to focus on the text input area so that
users can either continue with text input or hit enter to submit.

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
fix lint

sfc-gh-pchiu committed:
more fix

sfc-gh-pchiu committed:
fix hardcoded value

sfc-gh-pchiu committed:
more fixes

sfc-gh-pchiu committed:
change to get by id

sfc-gh-pchiu committed:
fix ChatInput.test.tsx

sfc-gh-pchiu committed:
fix import

sfc-gh-pchiu committed:
missing comma after merge

sfc-gh-pchiu committed:
fix vi after rebase

sfc-gh-pchiu committed:
chat py unittest (#9970)

## Describe your changes

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kajarenc committed:
Small improvements (#9982)

## Describe your changes
* Add overloads for better typing for chat_input
* better name _pop_upload_files to show that we also remove files from
storage
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Chat input with file uploader patch test (#9985)

## Describe your changes

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Add AcceptFile enum for proto (#9987)

## Describe your changes

Created a new enum for accept file values.
Added mapping of `chat_input` param to proto, and the mapping of proto
value to ts enum


## GitHub Issue Link (if applicable)

## Testing Plan

Locally tested on `chat_input` widget

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Clean and refactor ChatInput component (#10105)

## Describe your changes

Cleaned up `ChatInput.tsx` so that we will not be flagged by CI due to
`console.log`s.
I also added error messages for rejects files + refactored a number of
minor things.



## GitHub Issue Link (if applicable)

## Testing Plan

- Unit Tests (JS and/or Python)

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
`chat_test.py` improvements with parameterized and clean ups (#10054)

## Describe your changes

* Consolidated tests with `parameterized` as it's already adopted for
some test cases here
* Remove placeholder whenever not validated with expected values

## GitHub Issue Link (if applicable)

## Testing Plan

- pytests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Adopt new normalize file type API for chat input (#10053)

## Describe your changes

* Used the new normalize file type API introduced in #9986
* Cleaned up tests covered by `file_uploader_utils_test.py`


## GitHub Issue Link (if applicable)

## Testing Plan

- pytests

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Support rendering ChatInputValue as dictionary in st.write (#10129)

## Describe your changes
Render `ChatInputValue` as a proper dict in `st.write` by making it
inherits `Mapping` and implementing `to_dict` method. `st.write` could
then render it as
[custom_dict](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/elements/write.py#L484).

#### Before
<img width="517" alt="Screenshot 2025-01-08 at 6 08 50 PM"
src="https://github.com/user-attachments/assets/c95ce322-86f3-4367-ada8-b3e8533270b0"
/>

#### After
<img width="513" alt="Screenshot 2025-01-08 at 6 03 39 PM"
src="https://github.com/user-attachments/assets/12b8b457-a82c-4e78-a10a-c3ae0e1b8f24"
/> |



## GitHub Issue Link (if applicable)

## Testing Plan

- Unit Tests (JS and/or Python)

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Merge branch 'develop' into chat-input-with-file-uploder-mvp

sfc-gh-pchiu committed:
make `update-snapshots` for st.chat_input  (#10185)

## Describe your changes

Not final snapshots -- but useful to prevent CI check failures when
working on branches

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
fix element paddings and margins (#10187)

## Describe your changes

When reviewing the design I noticed that the spacings between elements
inside the chat input are off.
3 changes were made:
1. spacing between file icon and divider 12px > 8px (used negative
margin because the BaseUI icon padding are fixed by size)
2. spacing between divider and placeholder 14px > 12px
3. left padding for non-file enabled chat input should be `lg`

Before
<img width="634" alt="Screenshot 2025-01-15 at 2 53 34 AM"
src="https://github.com/user-attachments/assets/6b7322c0-b441-41a4-a7f3-aa3091944739"
/>

After
<img width="635" alt="Screenshot 2025-01-15 at 2 52 44 AM"
src="https://github.com/user-attachments/assets/c9d9ade9-a25b-4de3-a5cb-f195f6c53807"
/>




## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Consolidate border radii value and remove redundant css attr (#10184)

## Describe your changes

1. Move border radii (xxxl) to const so we have 1 source of truth for
its value throughout this component. In the future when we want to
change it to consume advanced theming's config we only need to modify 1
const.
2. There are quite a few places where we set `backgroundColor:
theme.colors.transparent` but they did not change the rendered outcome,
or could be handled by setting `overflow: "hidden"` in their parent. So
I re-organized the css attributes in these parent-children element and
a) move the `focus` attr to `StyledChatInputContainer` and b) remove
redundant attr after adding `overflow: hidden`


## GitHub Issue Link (if applicable)

## Testing Plan

Manual tested that chat input w or w/o file upload behaves the same

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Merge branch 'develop' into chat-input-with-file-uploder-mvp

sfc-gh-pchiu committed:
rerun snapshots

sfc-gh-pchiu committed:
File upload drag and drop UI (#10189)

## Describe your changes

Added drag and drop UI:


Introduced `fileDragged` state and a `showDropzone` flag. In "drop"
mode, multiple components have the following changes
* ChatInputContainer - border and height changes
* Input button - expand to fill the entire chat input container to allow
larger dropzone
* Text area - removed when user is dragging a file

### Demo (observe the following events)
* chat input changes upon dragged target entering page
* chat input returns to normal upon target leaving page
* dropping target inside the input uploads the file
* chat input returns to normal form upon dropping target outside the
input


https://github.com/user-attachments/assets/23af6044-ff9b-4815-8b6b-02a32e5b0a62



## GitHub Issue Link (if applicable)

## Testing Plan

- E2E test todo

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Merge branch 'develop' into chat-input-with-file-uploder-mvp

sfc-gh-pchiu committed:
Simplify chat input spacing based on new design spec (#10240)

## Describe your changes

A few tweaks to consolidate chat input styling attr. 
This [design
file](https://www.figma.com/design/svukmRMf0N9yQzdv8f7sgO/Streamlit-Open-Source-design-system?node-id=2324-44829&m=dev)
enumerated all chat input states and will be our source of truth here.

## GitHub Issue Link (if applicable)

## Testing Plan

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
file list preview for chat input (#10235)

## Describe your changes

* Move away from using pagination to display selected files
* Support 3 states: uploaded, uploading, error
* New tooltip element to work around tooltip not being able respect
placement for target within an `overflow: auto` element

#### Design

<img width="917" alt="Screenshot 2025-01-23 at 3 52 35 PM"
src="https://github.com/user-attachments/assets/1edfa4ce-b2fc-463d-b1ce-ad7681a81645"
/>

#### Demo

<img width="1308" alt="Screenshot 2025-01-23 at 3 59 39 PM"
src="https://github.com/user-attachments/assets/0dc1ffe7-ae6b-451a-af63-59437905c351"
/>



https://github.com/user-attachments/assets/0bc2da1f-d56c-44cd-a960-9b3d91421b5d




## GitHub Issue Link (if applicable)

## Testing Plan

- E2E testing to come

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
revert changes to UploadedFile(s)

sfc-gh-pchiu committed:
Merge branch 'develop' into chat-input-with-file-uploder-mvp

sfc-gh-pchiu committed:
add italic label for upload button, bolded drag and drop text (#10243)

## Describe your changes

1. Implemented upload button tooltip, added an `italicLabel` prop to
`StreamlitMarkdown`

<img width="544" alt="image"
src="https://github.com/user-attachments/assets/4c62cdfe-05e6-4fbf-b7ce-6d21073740f4"
/>

2. Updated font weight to 600 (700 in Figma but we typically reserve 700
for h1 header)
 
<img width="741" alt="image"
src="https://github.com/user-attachments/assets/eb937412-c57e-49ea-bbb4-2636876cf7db"
/>




## GitHub Issue Link (if applicable)

## Testing Plan

- Unit Tests (JS and/or Python)

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Revert "add italic label for upload button, bolded drag and drop text" (#10250)

Reverts streamlit/streamlit#10243

kajarenc committed:
Fix file_uploader in chat_input backend merge (#10252)

## Describe your changes

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
dropzone font weight

sfc-gh-pchiu committed:
Chat input with file uploader e2e tests (#10261)

## Describe your changes

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Merge branch 'develop' into chat-input-with-file-uploder-mvp

sfc-gh-pchiu committed:
fix import

sfc-gh-pchiu committed:
import fix  (#10263)

## Describe your changes

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-bnisco commented on frontend/lib/src/WidgetStateManager.ts:
```diff
@@ -318,6 +320,19 @@ export class WidgetStateManager {
     return this.setTriggerValueAtEndOfEventLoop(widget, source, fragmentId)
   }
 
+  public setChatInputValue(
```
**question:** `WidgetStateManager` is something that is used throughout the app, and thus should be as generic as possible. Is there a particular reason to keep this function (which is tied to chat) in this global class, rather than in some helper function instead?

sfc-gh-pchiu replied:
@kajarenc wdyt? I believe we introduced this function here looking to replace `setStringTriggerValue` and need to access the private functions here (`create/deleteWidgetState`, `onWidgetValueChanged`.) Could we move it out and expose these private functions?

kajarenc replied:
I trust your best judgements here. 

My understanding of context: `WidgetStateManager` should be able to work/handle all type of values produced by our widgets, and for each type of value we have corresponding function, like `setIntValue`, `setStringValue` etc. 

For chat_input we previously used `setStringTriggerValue` function that also lives in the `WidgetStateManager`
In this change we introduce a new type of value `ChatInputValue` (that contains information about files too rather than just text) and corresponding `setChatInputValue` should be part of `WidgetStateManager`. 

We plan to remove old `StringTriggerValue` and corresponding functions etc. as a follow-up of this project, we don't do it as part of this PR, because it is connected to removing corresponding proto fields, and it is big enough change to do it separately. 

Again, I don't have strong opinion here, so please follow your best judgement.

sfc-gh-bnisco replied:
Thank you for your comments! If my original question here requires a lot of refactoring, then feel free to not do it. I'd rather not cause a lot of unnecessary noise in this PR.

sfc-gh-pchiu replied:
Added this to our tech debt DB

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.test.tsx:
```diff
@@ -43,9 +49,39 @@ const getProps = (
     sendRerunBackMsg: vi.fn(),
     formsDataChanged: vi.fn(),
   }),
+  // @ts-expect-error
```
**question:** I understand this is test code, but is there a particular reason for this? Are there other approaches to working with mocks in typescript that can help the type system help us proactively identify issues when they may arise?

sfc-gh-pchiu replied:
I did a quick look into this -- looks like the error requires a bit more involved fix and we're currently silencing the same error in [FileUploader.text.tsx](https://github.com/streamlit/streamlit/blob/develop/frontend/lib/src/components/widgets/FileUploader/FileUploader.test.tsx#L75) and [CameraInput.test.tsx](https://github.com/streamlit/streamlit/blob/develop/frontend/lib/src/components/widgets/CameraInput/CameraInput.test.tsx#L54) 😕
I'll probably come back to this after addressing other issues (and hopefully fix those other tests, too)

sfc-gh-bnisco replied:
Thanks for providing other areas where we're already doing this exact pattern. Probably not worth the extra time, so feel free to leave it as-is for this PR!

sfc-gh-pchiu replied:
Added this to tech debt DB

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.test.tsx:
```diff
@@ -43,9 +49,39 @@ const getProps = (
 
+const mockChatInputValue = (
+  text: string,
+  _: UploadFileInfo[] = [] // TODO (sfc-gh-pchiu): implement mock for files
```
**question:** Do we need to address this TODO in this PR?

sfc-gh-pchiu replied:
Okay I now recall that in my discussion with @kajarenc we wanted to test uploading via playwright so this can be removed.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.test.tsx:
```diff
@@ -260,7 +296,7 @@ describe("ChatInput widget", () => {
     expect(chatInput).toBeDisabled()
 
-    const button = screen.getByRole("button")
+    const button = screen.getByTestId("stChatInputSubmitButton")
```
**suggestion:** We should utilize accessible selectors (such as `getByRole`) as often as possible, and only falling back to `getByTestId` as a last resort of no other methods are possible [React Testing Library docs here](https://testing-library.com/docs/queries/about#priority). Can we bring this back to an accessible selector perhaps with some modifications so that we're sure it selects the correct button?

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -22,55 +22,414 @@ import React, {
+          type: "error",
+          errorMessage: rejected.errors
+            .map(err => err.message)
+            .filter(err => err !== "")
```
**question:** Should this be `.filter(err => !!err)` to guard against null and undefined values too? Or is it guaranteed to be a string?

sfc-gh-pchiu replied:
It's guaranteed to be a string I think -- but this was carried over from a refactor so I didn't intentionally look to use that check. Changing to `.filter(err => !!err)` sounds good.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -22,55 +22,414 @@ import React, {
+    uploadClient
+      .uploadFile(
+        {
+          formId: "", // TODO[kajarnec] fix this probably with uploadFile refactoring
```
**question:** Do we need to address this TODO in this PR?

kajarenc replied:
As a big follow-up, I see a refactoring fronted (and maybe backend) of "file uploader" like widgets, rewriting class-based ones as function-based components, e.g., `file_uploader` and `st.camera_input`, having better API and more shared functionality moved to another component for all function-based ones, and cleaning up the comments like [this](https://github.com/streamlit/streamlit/blob/develop/frontend/lib/src/util/uploadFiles.ts#L79) or above one. 

I left this comment to have a good candidate for future refactoring; I think it should be done as a separate PR.

sfc-gh-bnisco replied:
@kajarenc Thanks for the context! Can you file file an internal tech debt item so we can keep track of this work?

sfc-gh-pchiu replied:
Added this to tech debt DB

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -22,55 +22,414 @@ import React, {
+  }: CreateUploadFileParams) =>
+  (fileURLs: IFileURLs, file: File): void => {
+    // Create an UploadFileInfo for this file and add it to our state.
+    const cancelToken = axios.CancelToken.source()
```
**suggestion:** `CancelToken` is [a deprecated API](https://axios-http.com/docs/cancellation) since it predates the [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController). Can we utilize AbortController instead so we're utilizing up-to-date APIs?

sfc-gh-pchiu replied:
It looks like this is used in `st.file_uploader` as well so it's probably safer to deprecate it separately. 
Added this to tech debt DB.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -142,9 +507,13 @@ function ChatInput({
-  }, [element])
+  }, [element, files])
+
+  useEffect(() => {
+    setDirty(isDirty(value, files))
+  }, [files, value])
```
**suggestion:** `dirty` should be refactored from a piece of state to a `useMemo` so we don't have an unnecessary `useState`/`useEffect` calling pattern. See the [React Docs on this](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes).

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -22,55 +22,414 @@ import React, {
+  const [fileDragged, setFileDragged] = useState(false)
+
+  const acceptFile = chatInputAcceptFileProtoValueToEnum(element.acceptFile)
+  const addFiles = (filesToAdd: UploadFileInfo[]): void => {
+    setFiles(currentFiles => [...currentFiles, ...filesToAdd])
+  }
```
**suggestion:** This should be put in a `useCallback` to prevent unnecessary re-renders

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -22,55 +22,414 @@ import React, {
+    setFiles(currentFiles => [...currentFiles, ...filesToAdd])
+  }
+
+  const isDirty = (value: string, files: UploadFileInfo[]): boolean => {
```
**suggestion:** This should either be a `useCallback` or moved into a standalone helper function (because this doesn't use local component-level state) to prevent unnecessary re-renders

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -22,55 +22,414 @@ import React, {
+  }
+
+  const isDirty = (value: string, files: UploadFileInfo[]): boolean => {
+    // TODO [kajarnec] add explanatory comment here.
```
**question:** Do we need to address this TODO in this PR?

kajarenc replied:
yeah, this "isDirty" was an add-on to existing  `[dirty, setDirty]` state to work with files too
It would be great if this could be refactored to use more react way primitives to address same logic.

sfc-gh-pchiu replied:
Converted this to a hook
```
useMemo(() => {
    if (files.some(f => f.status.type === "uploading")) {
      setDirty(false)
    }
    setDirty(value !== "" || files.length > 0)
  }, [files, value])
```
so that we don't need the `isDirty` calls where `files` or `value` change values

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -22,55 +22,414 @@ import React, {
+    return value !== "" || files.length > 0
+  }
+
+  const updateFile = (
+    id: number,
+    fileInfo: UploadFileInfo,
+    currentFiles: UploadFileInfo[]
+  ): UploadFileInfo[] => currentFiles.map(f => (f.id === id ? fileInfo : f))
+
+  const getFile = (
+    localFileId: number,
+    currentFiles: UploadFileInfo[]
+  ): UploadFileInfo | undefined => currentFiles.find(f => f.id === localFileId)
```
**suggestion:** Move these into standalone helper functions rather than in the component body to prevent unnecessary re-renders

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -22,55 +22,414 @@ import React, {
+    currentFiles: UploadFileInfo[]
+  ): UploadFileInfo | undefined => currentFiles.find(f => f.id === localFileId)
+
+  const deleteFile = (fileId: number): void => {
```
**suggestion:** Make this a `useCallback`

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/UploadedFile/ChatUploadedFile.tsx:
```diff
@@ -0,0 +1,110 @@
+    case "uploaded":
+      return <Icon content={InsertDriveFile} size="lg" />
+    default:
+      return null
```
**suggestion:** To get proactive type-time exhaustive checks, utilize `assertNever(type)` here. An example of this pattern can be [found here](https://github.com/streamlit/streamlit/blob/develop/frontend/lib/src/components/elements/DeckGlJsonChart/DeckGlJsonChart.tsx#L177)

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/UploadedFile/ChatUploadedFile.tsx:
```diff
@@ -0,0 +1,110 @@
+
+export const ChatUploadedFileIcon = ({
+  fileInfo,
+}: ChatUploadedFileIconProps): React.ReactElement | null => {
```
**suggestion (non-blocking):** Utilizing `export const ChatUploadedFileIcon: FC<ChatUploadedFileIconProps>` is terser than having to define both input and output types. The reason I'm mentioning this is that the return type of `React.ReactElement | null` is often better written as `React.ReactNode` since it encompasses other valid return types. Using `FC` will do that for you as well, which is my personal preference, but am totally fine with either!

[This article](https://dev.to/itswillt/explaining-reacts-types-940) has a nice venn diagram I often refer back to for the different types.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/UploadedFile/ChatUploadedFiles.tsx:
```diff
@@ -0,0 +1,46 @@
+export interface Props {
+  items: UploadFileInfo[]
+  onDelete: (id: number) => void
+  style?: React.CSSProperties
```
**suggestion:** Remove this, I don't see `style` prop being used in this component at all.

sfc-gh-bnisco commented on frontend/lib/src/util/utils.ts:
```diff
@@ -402,6 +403,27 @@ export function labelVisibilityProtoValueToEnum(
+      return AcceptFileValue.Single
+    case ChatInputProto.AcceptFile.MULTIPLE:
+      return AcceptFileValue.Multiple
+    default:
```
**suggestion:** Similar comment to above, utilize `assertNever(value)` here to get proactive exhaustive type checking

sfc-gh-bnisco commented on lib/streamlit/elements/widgets/chat.py:
```diff
@@ -42,12 +54,43 @@
+        return vars(self)
+
+
+TYPE_PAIRS = [
```
**suggestion:** Delete this, I don't see this being used anywhere

sfc-gh-bnisco commented on lib/streamlit/elements/widgets/chat.py:
```diff
@@ -267,9 +389,17 @@ def chat_input(
         disabled : bool
             Whether the chat input should be disabled. Defaults to ``False``.
 
+        file_type : str or list[str] or None
```
**question:** Is this a necessary feature? Extensions can be trivially modified, so I'm curious what the intent is behind providing this? If we are trying to determine actual type of uploaded files then checking mime types is a more reliable way of going about it.

kajarenc replied:
Yeah, the idea here is that we want to give users the ability to specify what type of files they want to allow to upload (either via `st.file_uploader` or file uploader in `st.chat_input`), e.g., only `jpg` and `gifs`, or let's say `pdfs`. 

Then, the front end handles that information during the file picking by allowing the upload of only files with the correct extension in the upload interaction. 

I absolutely agree that we can do better for checking based on mime type, etc., but for now, this feature is just for improving user experience, and deep checks are delegated to the streamlit app developer if needed in the app if needed.

lukasmasuch replied:
btw. I think we should also just support mimetypes here as well - basically, everything that is supported in the file input accept: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/accept#unique_file_type_specifiers

The OS also optimized uploads based on mime types, e.g. if you configure `images/*` it will provide access to the photo gallery / camera. @jrieke has started implementing a step towards adding that to `st.file_uploader` in https://github.com/streamlit/streamlit/pull/10068. Maybe we can merge this soon and add it to the chat input upload as well.

sfc-gh-bnisco commented on lib/tests/streamlit/elements/chat_test.py:
```diff
@@ -42,107 +52,60 @@ def test_nesting_is_disallowed(self):
+        [
+            ("👋", {"avatar": "👋", "type": BlockProto.ChatMessage.AvatarType.EMOJI}),
+            (
+                "https://static.streamlit.io/examples/cat.jpg",
```
**suggestion:** Tests shouldn't utilize remote resources since it can lead to non-determinism when running the test suite. Is there a local jpg we can utilize to test this?

sfc-gh-pchiu replied:
There are some pngs under `lib/tests/streamlit/commands` but am not sure if we should use them? cc @lukasmasuch 
Regardless, I'm thinking we may want to do this is a separate PR.

lukasmasuch replied:
It doesn't really matter in this case since these unit tests won't request the linked URL. The frontend would request the image, but that isn't involved in the unit tests. So, you can have an arbitrary URL here as long as it looks like a valid URL.

sfc-gh-pchiu replied:
Gotcha. I've changed it to a mock url to avoid confusion since we're really matching strings here for `avatar`.

sfc-gh-pchiu committed:
update filesize text to normal size rather than small

sfc-gh-pchiu committed:
opt for getByRole in test

sfc-gh-pchiu committed:
address comments

sfc-gh-pchiu committed:
revert pytest changes

sfc-gh-pchiu committed:
don't use actual url

sfc-gh-lmasuch commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -17,60 +17,431 @@
+  theme: EmotionTheme
+}
+
+const FileUploadArea = ({
```
nit: I think some of the file handling logic e.g. `FileUploadArea` and `createUploadFileHandler` could be moved to a dedicated file.

lukasmasuch commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -17,60 +17,431 @@
+  theme: EmotionTheme
+}
+
+const FileUploadArea = ({
```
I think some of the file handling logic e.g. `FileUploadArea` and `createUploadFileHandler` could be moved to a dedicated file to keep the main ChatInput component leaner.

sfc-gh-pchiu replied:
Refactored in #10291

lukasmasuch commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -17,60 +17,431 @@
+}: FileUploadAreaProps): React.ReactElement =>
+  showDropzone ? (
+    <StyledFileUploadDropzone {...getRootProps()}>
+      <input {...getInputProps()} />
+      Drag and drop files here
+    </StyledFileUploadDropzone>
+  ) : (
+    <>
+      <div data-testid="stChatInputFileUploadButton" {...getRootProps()}>
+        <input {...getInputProps()} />
+        <BaseButton kind={BaseButtonKind.MINIMAL} disabled={disabled}>
```
I'm not sure what, but something seems to be broken with the `input` since the `accept` parameter is not correctly processed for the file chooser. The same seems to work fine for `file_uploader`

lukasmasuch replied:
What's interesting: it seems to be only broken in Chrome, but works fine in Safari and Firefox.

 My first suggestion would be to update the react-dropzone dependency. We are currently using react-dropzone==12.1.0 which is over 3 years old, and you can find lots of fixed bugs about accept in the react-dorpzone repo.

sfc-gh-pchiu replied:
#10284 to update the dependency

sfc-gh-pchiu replied:
The `accept` param seems to be working now in Chrome with some testing using `st.chat_input(file_type=...` after upgrading react-dropzone and refactoring with custom mimetype #10292

lukasmasuch replied:
Awesome, it also works for me [in the demo](https://file-uploader-in-chat-input.streamlit.app/) now.

sfc-gh-pchiu committed:
remove unused param

sfc-gh-pchiu committed:
fix accept

sfc-gh-pchiu committed:
Merge branch 'develop' into chat-input-with-file-uploder-mvp

sfc-gh-pchiu committed:
Split `FileUploadArea` and handlers out from `ChatInput` (#10291)

## Describe your changes

as title -- refactored to keep `ChatInput` leaner

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
Merge branch 'develop' into chat-input-with-file-uploder-mvp

sfc-gh-pchiu committed:
use new mimetype utils for dropzone `accept` value (#10292)

## Describe your changes

Also adopt the util function added in #10288 for ChatInput 

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
use FilePayload object & fix timeout for filechooser (#10266)

## Describe your changes

Followed #10267 to adopt FilePayload objects in Playwright tests
Also fixed `filechooser` hitting timeouts in some e2e test runs.

## GitHub Issue Link (if applicable)

## Testing Plan

- E2E Tests


---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
break file upload tests out since st_chat_input_test was flaky with too many chat inputs

sfc-gh-pchiu committed:
fix

sfc-gh-pchiu committed:
fix

sfc-gh-pchiu committed:
revert unnecessary change

sfc-gh-pchiu committed:
remove extra test

sfc-gh-pchiu committed:
snaps

sfc-gh-pchiu committed:
File upload for chat button tooltip (#10269)

## Describe your changes

Added tooltip for file upload button

<img width="301" alt="image"
src="https://github.com/user-attachments/assets/eb32d681-f813-4279-b0cf-71905c3ba5fb"
/>

Design:
<img width="355" alt="image"
src="https://github.com/user-attachments/assets/d6b1680e-fcfa-43a2-a9de-d4c2e26d906b"
/>



## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

sfc-gh-pchiu committed:
fix hover in snaps

sfc-gh-pchiu committed:
address comments

sfc-gh-pchiu committed:
Merge branch 'develop' into chat-input-with-file-uploder-mvp

sfc-gh-bnisco commented on e2e_playwright/st_chat_input_file_upload.py:
```diff
@@ -0,0 +1,21 @@
+
+import streamlit as st
+
+v1 = st.container().chat_input("Chat input 1 (single file)", accept_file=True)
```
**note:** It has been the practice so far to colocate tests into a single file due to the perf lifecycle of having to spin up/down a new streamlit app when this runs in CI. If there isn't a technical reason why this should be split out into its own test file, the pattern so far would be to put this in the existing `st_chat_input.py`/`st_chat_input_test.py` files

sfc-gh-pchiu replied:
I split these out as I observed some snapshots failed due to chat input boxes overlapping/being blocked -- e.g. snapshot-tests-failures from [this run](https://github.com/streamlit/streamlit/actions/runs/13045084598). And I figured it was due to `st_chat_input.py` being too crammed with input boxes. 

But I also wouldn't like impacting perf with a new app. I'll try expanding the viewport height of `st_chat_input.py` and see if that fixes the overlapping

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -129,11 +309,18 @@ function ChatInput({
     setScrollHeight(getScrollHeight())
   }
 
+  useMemo(() => {
+    // Disable send button if there are files still being uploaded
+    if (files.some(f => f.status.type === "uploading")) {
+      setDirty(false)
+    }
+    setDirty(value !== "" || files.length > 0)
+  }, [files, value])
```
**suggestion:** the function given to a `useMemo` needs to be a pure function [React docs](https://react.dev/reference/react/useMemo#parameters), but this has side-effects of setting state. Swapping this out for a `useEffect` should fix this!

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/fileUpload/ChatUploadedFile.tsx:
```diff
@@ -0,0 +1,120 @@
+export const ChatUploadedFileIcon: FC<ChatUploadedFileIconProps> = ({
+  fileInfo,
+}) => {
+  const fileStatus: FileStatus["type"] = fileInfo.status.type
```
**question:** Is this explicit type annotation necessary? Seems like `fileInfo` has a proper type attached to it, so I would expect the following to work:

```tsx
const { type: fileStatus } = fileInfo.status
```

sfc-gh-pchiu committed:
unpack type

sfc-gh-pchiu committed:
Merge branch 'develop' into chat-input-with-file-uploder-mvp

sfc-gh-pchiu committed:
address comment, move e2e tests back to st_chat_input n set height

sfc-gh-pchiu committed:
take away hover & snaps

sfc-gh-bnisco reviewed (approved):
Thank you for the iterations! Sounds like there are a number of tracked follow-up improvements to be made, with mime types being probably the most impactful one from a UX standpoint. I'm excited about this work and getting it into the hands of users!

sfc-gh-pchiu committed:
fix setDirty hook bug

sfc-gh-pchiu committed:
rearrange params

sfc-gh-pchiu committed:
only allow files to trigger dropzone

sfc-gh-pchiu committed:
make ChatInputValue mutable

sfc-gh-pchiu commented:
Updates
1. Addressed product feedback
    1. moved `file_type` param to right after `accept_file`
    2. showed dropzone only when files are dragged (exclude images, links from webpage)
    3. allowed return value to be mutable (e.g. can assign/delete keys)
2. Fixed [a bug](https://github.com/streamlit/streamlit/pull/9491/commits/587269bb083346ffe4a21fb9544de452362fa44e#diff-559594d48f19e6ea4c58bfa5cb445c7f4e4596c3865e7e88b7f0af13abf146eaR316-R322) introduced when refactoring `isDirty` function
3. (new) Addressed design feedback: allow vertical divider to grow when input area height is extended

sfc-gh-pchiu committed:
address design feedback on vert line

sfc-gh-pchiu committed:
Merge branch 'develop' into chat-input-with-file-uploder-mvp

sfc-gh-pchiu committed:
implemented design feedback for vert divider, move a few FileUploadArea styles to subdir, add playwright

sfc-gh-pchiu committed:
spacing nit, new snaps

sfc-gh-pchiu committed:
snaps

sfc-gh-pchiu committed:
Merge branch 'develop' into chat-input-with-file-uploder-mvp

lix8886 commented:
大佬再支持一下提供语音输入！！
Could you kindly consider adding **voice input** support?
